In [ ]:

import pandas as pd
x = [1,2,2,3,4,4,5,6]
y = [2,3,4,4,4,6,5,7]

d = {"y":y, "x":x}
df = pd.DataFrame(d)
df

y  x
0  2  1
1  3  2
2  4  2
3  4  3
4  4  4
5  6  4
6  5  5
7  7  6

In [ ]:
import numpy as np
df2=df.copy()
df2["y"]=np.ones(len(df))
df2

y  x
0  1.0  1
1  1.0  2
2  1.0  2
3  1.0  3
4  1.0  4
5  1.0  4
6  1.0  5
7  1.0  6

In [ ]:
X=np.array(df2)
X

array([[1., 1.],
       [1., 2.],
       [1., 2.],
       [1., 3.],
       [1., 4.],
       [1., 4.],
       [1., 5.],
       [1., 6.]])

In [ ]:
Xt = X.T
Xt

array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 2., 2., 3., 4., 4., 5., 6.]])

In [ ]:
M=Xt@X
M

array([[  8.,  27.],
       [ 27., 111.]])

In [ ]:
y=np.array(df[["y"]])
y

array([[2],
       [3],
       [4],
       [4],
       [4],
       [6],
       [5],
       [7]])

In [ ]:
detM = np.linalg.det(M)
detM

np.float64(159.0)

In [ ]:
inversa=np.linalg.inv(M)
inversa


array([[ 0.69811321, -0.16981132],
       [-0.16981132,  0.05031447]])

In [ ]:
pinversa=np.linalg.pinv(M)
pinversa

array([[ 0.69811321, -0.16981132],
       [-0.16981132,  0.05031447]])

In [ ]:
betas = inversa @ Xt @ y
betas

array([[1.50943396],
       [0.8490566 ]])

# Mas rapido:

Ver este link relacionado con la matriz pseudo inversa de Moore-Penrose:

https://chatgpt.com/share/68c0df00-15d8-800f-9b6b-61b0a212a247

In [ ]:
import pandas as pd
x = [1,2,2,3,4,4,5,6]
y = [2,3,4,4,4,6,5,7]

d = {"x":x, "y":y}
df = pd.DataFrame(d)
df

x  y
0  1  2
1  2  3
2  2  4
3  3  4
4  4  4
5  4  6
6  5  5
7  6  7

In [ ]:
df2=df.copy()
df2["y"]=np.ones(len(df))
df2

x    y
0  1  1.0
1  2  1.0
2  2  1.0
3  3  1.0
4  4  1.0
5  4  1.0
6  5  1.0
7  6  1.0

In [ ]:
X=np.array(df2)
X.shape

(8, 2)

In [ ]:
Xpinv= np.linalg.pinv(X)
Xpinv.shape

(2, 8)

In [ ]:
y = np.array(df[["y"]])
y.shape

(8, 1)

In [ ]:
betas = Xpinv @ y
print(betas.shape)
betas

(2, 1)


array([[0.8490566 ],
       [1.50943396]])

# Ojo la pseudo inversa de Moore-Penrose ya incluye todo (matrices no tienen que ser cuadradas y el det(AtA) puede ser cero)
(la inversa no lo hace directamente pues la matriz debe ser cuadrada, y si fuera cuadrada su det debe ser distinto de cero, ver mas abajo):

In [ ]:
Xinv= np.linalg.inv(X)
Xinv.shape

LinAlgError: Last 2 dimensions of the array must be square

# Cálculo de Errores (manualmente)

In [ ]:
df["ypred"] = X @ betas
df

x  y     ypred
0  1  2  2.358491
1  2  3  3.207547
2  2  4  3.207547
3  3  4  4.056604
4  4  4  4.905660
5  4  6  4.905660
6  5  5  5.754717
7  6  7  6.603774

In [ ]:
# error = y - ypred
df["y-ypred"] =  df["y"] - df["ypred"]
df

x  y     ypred   y-ypred
0  1  2  2.358491 -0.358491
1  2  3  3.207547 -0.207547
2  2  4  3.207547  0.792453
3  3  4  4.056604 -0.056604
4  4  4  4.905660 -0.905660
5  4  6  4.905660  1.094340
6  5  5  5.754717 -0.754717
7  6  7  6.603774  0.396226

In [ ]:
# error = y - ypred
df["error^2"] = df["y-ypred"]**2
df

x  y     ypred   y-ypred   error^2
0  1  2  2.358491 -0.358491  0.128515
1  2  3  3.207547 -0.207547  0.043076
2  2  4  3.207547  0.792453  0.627981
3  3  4  4.056604 -0.056604  0.003204
4  4  4  4.905660 -0.905660  0.820221
5  4  6  4.905660  1.094340  1.197579
6  5  5  5.754717 -0.754717  0.569598
7  6  7  6.603774  0.396226  0.156995

In [ ]:
error_cuad_total = df["error^2"].sum()
error_cuad_total

np.float64(3.5471698113207553)

In [ ]:
# MSE = error_cuad_medio
mse = error_cuad_total / len(df)
mse

np.float64(0.4433962264150944)

In [ ]:
rmse = mse**(1/2)
rmse

# Histograma de errores

In [ ]:
df["y-ypred"].hist()

KeyError: 'y-ypred'

# Actividad de consulta práctica hasta las 7:40 pm:

1. Calcular el error absoluto total y el MAE (error absoluto medio). Ayuda: usar el df y extenderlo como se hizo para el cuadratico.

2. MAPE

3. Otras métricas: R2, entre otras.

*Recuerden:

Replicar lo anterior en cada dataset (el del video de youtube, el de las ventas y reg mutiple del link donde está la formula del OLS, y si alcanzan pueden ir replicando en su datset del caso de estudio)


In [ ]:
# Calculate predicted values
df["ypred"] = X @ betas

# Calculate error
df["y-ypred"] =  df["y"] - df["ypred"]

# Calculate absolute error
df["abs_error"] = abs(df["y-ypred"])
display(df)

# Calculate total absolute error
total_abs_error = df["abs_error"].sum()
print(f"Total Absolute Error: {total_abs_error}")

# Calculate Mean Absolute Error (MAE)
mae = total_abs_error / len(df)
print(f"Mean Absolute Error (MAE): {mae}")

# Calculate Absolute Percentage Error
df["abs_perc_error"] = (df["abs_error"] / df["y"]) * 100
display(df)

# Calculate Mean Absolute Percentage Error (MAPE)
mape = df["abs_perc_error"].mean()
print(f"Mean Absolute Percentage Error (MAPE): {mape}%")

# Calculate squared error
df["error^2"] = df["y-ypred"]**2

# Calculate R2
# Calculate the mean of y
y_mean = df["y"].mean()

# Calculate the total sum of squares (SST)
df["y-y_mean"] = df["y"] - y_mean
df["(y-y_mean)^2"] = df["y-y_mean"]**2
sst = df["(y-y_mean)^2"].sum()

# The sum of squared errors (SSE) is already calculated as error_cuad_total
sse = df["error^2"].sum()

# Calculate R2
r2 = 1 - (sse / sst)
print(f"R2 Score: {r2}")
display(df)

x  y     ypred   y-ypred   error^2  abs_error  abs_perc_error  y-y_mean  \
0  1  2  2.358491 -0.358491  0.128515   0.358491       17.924528    -2.375   
1  2  3  3.207547 -0.207547  0.043076   0.207547        6.918239    -1.375   
2  2  4  3.207547  0.792453  0.627981   0.792453       19.811321    -0.375   
3  3  4  4.056604 -0.056604  0.003204   0.056604        1.415094    -0.375   
4  4  4  4.905660 -0.905660  0.820221   0.905660       22.641509    -0.375   
5  4  6  4.905660  1.094340  1.197579   1.094340       18.238994     1.625   
6  5  5  5.754717 -0.754717  0.569598   0.754717       15.094340     0.625   
7  6  7  6.603774  0.396226  0.156995   0.396226        5.660377     2.625   

   (y-y_mean)^2  
0      5.640625  
1      1.890625  
2      0.140625  
3      0.140625  
4      0.140625  
5      2.640625  
6      0.390625  
7      6.890625

Total Absolute Error: 4.56603773584906
Mean Absolute Error (MAE): 0.5707547169811324


x  y     ypred   y-ypred   error^2  abs_error  abs_perc_error  y-y_mean  \
0  1  2  2.358491 -0.358491  0.128515   0.358491       17.924528    -2.375   
1  2  3  3.207547 -0.207547  0.043076   0.207547        6.918239    -1.375   
2  2  4  3.207547  0.792453  0.627981   0.792453       19.811321    -0.375   
3  3  4  4.056604 -0.056604  0.003204   0.056604        1.415094    -0.375   
4  4  4  4.905660 -0.905660  0.820221   0.905660       22.641509    -0.375   
5  4  6  4.905660  1.094340  1.197579   1.094340       18.238994     1.625   
6  5  5  5.754717 -0.754717  0.569598   0.754717       15.094340     0.625   
7  6  7  6.603774  0.396226  0.156995   0.396226        5.660377     2.625   

   (y-y_mean)^2  
0      5.640625  
1      1.890625  
2      0.140625  
3      0.140625  
4      0.140625  
5      2.640625  
6      0.390625  
7      6.890625

Mean Absolute Percentage Error (MAPE): 13.463050314465423%
R2 Score: 0.8015569336324053


x  y     ypred   y-ypred   error^2  abs_error  abs_perc_error  y-y_mean  \
0  1  2  2.358491 -0.358491  0.128515   0.358491       17.924528    -2.375   
1  2  3  3.207547 -0.207547  0.043076   0.207547        6.918239    -1.375   
2  2  4  3.207547  0.792453  0.627981   0.792453       19.811321    -0.375   
3  3  4  4.056604 -0.056604  0.003204   0.056604        1.415094    -0.375   
4  4  4  4.905660 -0.905660  0.820221   0.905660       22.641509    -0.375   
5  4  6  4.905660  1.094340  1.197579   1.094340       18.238994     1.625   
6  5  5  5.754717 -0.754717  0.569598   0.754717       15.094340     0.625   
7  6  7  6.603774  0.396226  0.156995   0.396226        5.660377     2.625   

   (y-y_mean)^2  
0      5.640625  
1      1.890625  
2      0.140625  
3      0.140625  
4      0.140625  
5      2.640625  
6      0.390625  
7      6.890625

## Análisis de Métricas de Error y Rendimiento del Modelo

En las celdas anteriores, calculamos varias métricas para evaluar el rendimiento de nuestro modelo de regresión. A continuación, se presenta un análisis de cada una de ellas en español:

**Error Cuadrático Total (Sum of Squared Errors - SSE):**

- **Concepto:** Es la suma de los cuadrados de las diferencias entre los valores reales de 'y' y los valores predichos por el modelo ('ypred').
- **Significado:** Representa la variabilidad total de los errores del modelo. Un valor menor indica que las predicciones del modelo están más cerca de los valores reales.
- **Representación en las cifras:** En este caso, el `error_cuad_total` es `3.5471698113207553`.

**Error Cuadrático Medio (Mean Squared Error - MSE):**

- **Concepto:** Es el promedio de los errores cuadráticos. Se calcula dividiendo el Error Cuadrático Total por el número de observaciones.
- **Significado:** Mide la dispersión promedio de los errores al cuadrado. Al igual que el SSE, un valor menor indica un mejor ajuste del modelo.
- **Representación en las cifras:** El `mse` es `0.4433962264150944`.

**Raíz del Error Cuadrático Medio (Root Mean Squared Error - RMSE):**

- **Concepto:** Es la raíz cuadrada del MSE.
- **Significado:** Proporciona una medida del error en las mismas unidades que la variable de respuesta ('y'). Es más interpretable que el MSE porque no está en unidades cuadradas. Un RMSE más bajo indica un mejor rendimiento del modelo.
- **Representación en las cifras:** El `rmse` es `0.6658800390574074`.

**Error Absoluto Total:**

- **Concepto:** Es la suma de los valores absolutos de las diferencias entre los valores reales de 'y' y los valores predichos por el modelo ('ypred').
- **Significado:** Representa la magnitud total de los errores sin considerar su dirección.
- **Representación en las cifras:** El `total_abs_error` es `4.56603773584906`.

**Error Absoluto Medio (Mean Absolute Error - MAE):**

- **Concepto:** Es el promedio de los errores absolutos. Se calcula dividiendo el Error Absoluto Total por el número de observaciones.
- **Significado:** Mide la magnitud promedio de los errores sin considerar su dirección. Es menos sensible a valores atípicos que el MSE o RMSE. Un MAE más bajo indica un mejor rendimiento del modelo.
- **Representación en las cifras:** El `mae` es `0.5707547169811324`.

**Error Porcentual Absoluto Medio (Mean Absolute Percentage Error - MAPE):**

- **Concepto:** Es el promedio de los errores porcentuales absolutos. Se calcula como la suma de los errores absolutos dividida por los valores reales de 'y', y luego multiplicada por 100.
- **Significado:** Mide el error promedio como un porcentaje de los valores reales. Es útil para comparar el rendimiento del modelo en diferentes conjuntos de datos o variables con diferentes escalas. Un MAPE más bajo indica un mejor rendimiento. Es importante tener precaución con este métrica si hay valores de 'y' cercanos a cero.
- **Representación en las cifras:** El `mape` es `13.463050314465423%`.

**R-cuadrado (R2 Score):**

- **Concepto:** También conocido como coeficiente de determinación. Mide la proporción de la varianza en la variable dependiente ('y') que es predecible a partir de la(s) variable(s) independiente(s) ('x') en un modelo de regresión.
- **Significado:** Indica qué tan bien se ajusta el modelo a los datos. Un valor de R2 de 1.0 significa que el modelo explica toda la variabilidad en 'y', mientras que un valor de 0.0 significa que el modelo no explica ninguna variabilidad. Un R2 más alto generalmente indica un mejor ajuste.
- **Representación en las cifras:** El `r2` es `0.8015569336324053`. Esto significa que aproximadamente el 80.16% de la variabilidad en 'y' es explicada por el modelo.

En resumen, estas métricas nos dan una idea cuantitativa de qué tan bien nuestro modelo de regresión lineal se ajusta a los datos. Un RMSE y MAE bajos, junto con un R2 alto, sugieren un buen rendimiento del modelo en este conjunto de datos.